examples/react-full-v1/pages/star-wars/index/+onBeforeRender.ts

```diff
@@ -1,11 +1,13 @@
 // https://vike.dev/onBeforeRender
 
-import type { Config, PageContextServer } from 'vike/types'
+import type { Config } from 'vike/types'
 import { filterMoviesData, getStarWarsMovies, getTitle } from './getStarWarsMovies'
 
-const onBeforeRender: Config['onBeforeRender'] = async (
-  pageContext: PageContextServer
-): Promise<{ pageContext: Partial<Vike.PageContext> }> => {
+type OnBeforeRender = Extract<Config['onBeforeRender'], Function>
+
+const onBeforeRender: OnBeforeRender = async (
+  pageContext: Parameters<OnBeforeRender>[0]
+): Extract<ReturnType<OnBeforeRender>, Promise<any>> => {
```

Comment:
My personal preference as a user would be to read just this

```
const onBeforeRender = async (
  pageContext: PageContextServer
): Promise<{ pageContext: Partial<Vike.PageContext> }> => {
```

Because if I need to modify the implementation of the hook, it tells me super clearly what's available in the parameters and what I need to return. For me, having the information I need in the code exactly where I need it has even more value than the actual safety provided by TypeScript.

But it's my personal taste so what about this: we provide various styles across examples:
* in some examples we have no explicit types (yes, not fully safe)
* in some examples we use only `Config['...']` and no other explicit types (yes, not fully safe)
* in some examples we use the style I prefer
* in some examples we use `Extract<>`

And then it's up to the user to use the style they like, if they prefer safety, or readability, or conciseness. I could even write documentation on vike.dev about this. What do you think?

Reply:
Ok that's a good point: from a readability perspective, it is indeed much better.

> documentation on vike.dev 

Agreed: seems paramount to me to share with the user the knowledge we found out in this PR.

I'm actually tempted to make the verbose version our default choice, as I think it's a very good point that the readability of it is very neat. Maybe I(/we) can mull over it a bit more and then take a decision, but let me know if you feel blocked.

Reply:
> I could even write

(👍)

Reply:
> * But maybe we can make it simpler by having Vike providing some utility type (I can further dig later this week if you want me to).

Yes, for having some kind of simpler `Extract<>`-like utilities. I'll dig into that too but again that wouldn't be as a user/reader my favorite style, because it doesn't tell me what comes in and out of the function in simple, straightforward terms. If that's your favorite style though, I still think it would be cool to show/document different styles

Reply:
> Maybe I(/we) can mull over it a bit more and then take a decision

no rush. I'll still want to push a few things to this PR regarding your other points, e.g. `PageContextServer | PageContextClient`, and if you need to think a bit more about it, take your time :pray: